### ISSUE TITLE
AssertionError with degenerate prediction boxes

### ISSUE THREAD

**Issue by wbl66** [Gorilla-Lab-SCUT/PaDT#22]
Thanks for your excellent work!
However, in the **generalized box_iou()** function, I encountered an **AssertionError** with the model's prediction box:
```
def generalized_box_iou(self, boxes1, boxes2):
        """
        Generalized IoU from https://giou.stanford.edu/

        The boxes should be in [x0, y0, x1, y1] format

        Returns a [N, M] pairwise matrix, where N = len(boxes1)
        and M = len(boxes2)
        """
        # degenerate boxes gives inf / nan results
        # so do an early check
        assert (boxes1[:, 2:] >= boxes1[:, :2]).all() # AssertionError
        assert (boxes2[:, 2:] >= boxes2[:, :2]).all()
        iou, union = self.box_iou(boxes1, boxes2)

        lt = torch.min(boxes1[:, None, :2], boxes2[:, :2])
        rb = torch.max(boxes1[:, None, 2:], boxes2[:, 2:])

        wh = (rb - lt).clamp(min=0)  # [N,M,2]
        area = wh[:, :, 0] * wh[:, :, 1]

        return iou - (area - union) / area, iou
```
Specifically, **boxes1** represents the predicted box provided by the model, so we cannot guarantee it will pass the assert. May I ask if we need to add a fix step for the predicted box as follows:
```
def fix_degenerate_boxes(boxes):
        # boxes: [N,4] in [x0,y0,x1,y1]
        
        x0 = boxes[:, 0].clamp(0, 1)
        y0 = boxes[:, 1].clamp(0, 1)
        x1 = boxes[:, 2].clamp(0, 1)
        y1 = boxes[:, 3].clamp(0, 1)

        # ensure x1>=x0 and y1>=y0
        x1 = torch.maximum(x1, x0 + 1e-6)
        y1 = torch.maximum(y1, y0 + 1e-6)

        return torch.stack([x0, y0, x1, y1], dim=-1)
```
I would greatly appreciate any suggestions!

**Comment by Dyb3438** [comment]
https://github.com/Gorilla-Lab-SCUT/PaDT/blob/main/src/PaDT/trainer/padt_sft_trainer.py#L521

The box predictions of the model follow the format: (cx, cy, w, h), where cx, cy, w, and h are projected using the Sigmoid function, with their values in the range [0, 1]. The boxes are represented as [x1, y1, x2, y2], where x1 = cx - w / 2 and x2 = cx + w / 2. Therefore, I believe the model's predictions should pass the assert check successfully.

Indeed, you could also add some clamp operations and epsilon to preprocess them. That's no problem.

**Comment by wbl66** [comment]
Thanks for your explanation. I will debug again, and once I solve the problem, I will close this issue.
